File: tlg/src/main/java/ch/bailu/tlg/StateRunning.java
package ch.bailu.tlg;

import java.util.Random;

public class StateRunning extends State {
	public static final int SHAPE_PER_LEVEL = 7;
	
	public static final int ID=3;
	public static final String NAME="Playing"; 
	
	public StateRunning(InternalContext c) {
		super(c);
		context.setStatusText(NAME);
	}

	
	public State init(PlatformContext c) {
		erase();
		initPreview(c);
		context.setStatusText(NAME);
		return initMovingShape(c);
	}
	
	
    public void erase() {
        context.previewMatrix.erase();
        context.mainMatrix.erase();
        context.currentScore.reset();
    }

    
    private void initPreview(PlatformContext gc) {
        Random random = new Random();
        
        int shape=random.nextInt(context.currentScore.getLevel() * SHAPE_PER_LEVEL)+1;
        int color=gc.getColor(shape%gc.countOfColor());
        int turn=random.nextInt(4);
        
        context.previewMatrix.erase();
        context.previewMatrix.setRandomMovingShape(shape, color, turn);
        context.previewMatrix.centerMovingShape();
    }

    private State initMovingShape(PlatformContext gc) {
        boolean r=initTarget();
        if (r) {
            initPreview(gc);
        } else {
            return State.gameOverStateFactory(context, gc);
        }

        return this;
    }
    
    
    
    private boolean initTarget() {
        return context.mainMatrix.setMovingShape(context.previewMatrix.getMovingShape());
    }

    
    @Override
    public State moveRight(PlatformContext c) {
    	context.mainMatrix.moveShapeRight();
    	return this;
    }

    @Override
    public State moveLeft(PlatformContext c) {
    	context.mainMatrix.moveShapeLeft();
    	return this;
    }
    
    @Override
    public State moveDown(PlatformContext c) {
        if (!context.mainMatrix.moveShapeDown()) {
            removeLines();
            return initMovingShape(c);
        }
    	return this;
    }


    @Override
    public State moveTurn(PlatformContext c) {
    	context.mainMatrix.moveShapeTurn();
    	return this;
    }

    
    private void removeLines() {
        context.currentScore.addLines(context.mainMatrix.eraseLines());
        context.setStatusText(NAME);
        
    }
    
    @Override
	public State togglePause(PlatformContext c) {
		return new StatePaused(context).init(c);
	}

	@Override
	public int getTimerInterval() {
		return context.currentScore.getTimerInterval();
	}


	@Override
	public int getID() {
		return ID;
	}

}

File: js/tlg-lib/src/state/StateRunning.ts
import { State } from './State'
import { InternalContext } from '../context/InternalContext'
import { PlatformContext } from '../context/PlatformContext'
import { HighscoreList } from '../tlg.score/HighscoreList'
import { StateID } from './StateID'
import { StorageContext } from '../context/StorageContext'

export class StateRunning extends State {
    public static readonly SHAPE_PER_LEVEL = 7

    public static readonly NAME = 'Playing'

    constructor (c:InternalContext, s:StorageContext) {
        super(c, s)
        this.context.setStatusText(StateRunning.NAME)
        console.log('StateRunning::constructor()')
    }

    public getID (): number {
        return StateID.ID_RUNNING
    }

    public init (c:PlatformContext):StateID {
        console.log('StateRunning::Init()')
        this.erase()
        this.initPreview(c)
        return this.initMovingShape(c)
    }

    public erase () {
        this.context.previewMatrix.erase()
        this.context.mainMatrix.erase()
        this.context.currentScore.reset()
    }

    private initPreview (gc:PlatformContext) {
        const shape = this.random(this.context.currentScore.getLevel() * StateRunning.SHAPE_PER_LEVEL) + 1
        const color = gc.getColor(shape % gc.countOfColor())
        const turn = this.random(4)

        this.context.previewMatrix.erase()
        this.context.previewMatrix.setRandomMovingShape(shape, color, turn)
        this.context.previewMatrix.centerMovingShape()
    }

    private initMovingShape (gc:PlatformContext):StateID {
        if (this.initTarget()) {
            this.initPreview(gc)
            return StateID.IGNORE
        }
        return this.gameOverState(gc)
    }

    private gameOverState (p:PlatformContext):StateID {
        const highscoreList = new HighscoreList(p)

        if (highscoreList.haveNewHighscore(this.context.currentScore.getScore())) {
            return new StateID(StateID.ID_HIGHSCORE).toReset()
        } else {
            return new StateID(StateID.ID_LOCKED).toReset()
        }
    }

    private initTarget ():boolean {
        return this.context.mainMatrix.setMovingShape(this.context.previewMatrix.getMovingShape())
    }

    public moveRight (c:PlatformContext) {
        this.context.mainMatrix.moveShapeRight()
        return StateID.IGNORE
    }

    public moveLeft (c:PlatformContext) {
        this.context.mainMatrix.moveShapeLeft()
        return StateID.IGNORE
    }

    public moveDown (c:PlatformContext):StateID {
        if (!this.context.mainMatrix.moveShapeDown()) {
            this.removeLines()
            return this.initMovingShape(c)
        }
        return StateID.IGNORE
    }

    public moveTurn (c:PlatformContext) {
        this.context.mainMatrix.moveShapeTurn()
        return StateID.IGNORE
    }

    private removeLines ():void {
        this.context.currentScore.addLines(this.context.mainMatrix.eraseLines())
        this.context.setStatusText(StateRunning.NAME)
    }

    public togglePause (c:PlatformContext):StateID {
        return new StateID(StateID.ID_PAUSED)
    }

    public getTimerInterval ():number {
        return this.context.currentScore.getTimerInterval()
    }

    public random (max:number):number {
        return Math.floor(Math.random() * max)
    }
}
